pageserver/src/basebackup.rs

```diff
@@ -133,6 +134,84 @@ where
     ctx: &'a RequestContext,
 }
 
+/// A sink that accepts SLRU blocks ordered by key and forwards
+/// full segments to the archive.
+struct SlruSegmentsBuilder<'a, 'b, W>
+where
+    W: AsyncWrite + Send + Sync + Unpin,
+{
+    ar: &'a mut Builder<&'b mut W>,
+    buf: Vec<u8>,
+    current_segment: Option<(SlruKind, u32)>,
+}
+
+impl<'a, 'b, W> SlruSegmentsBuilder<'a, 'b, W>
+where
+    W: AsyncWrite + Send + Sync + Unpin,
+{
+    fn new(ar: &'a mut Builder<&'b mut W>) -> Self {
+        Self {
+            ar,
+            buf: Vec::new(),
+            current_segment: None,
+        }
+    }
+
+    async fn add_block(&mut self, key: &Key, block: Bytes) -> anyhow::Result<()> {
+        let (kind, segno, _) = key_to_slru_block(*key)?;
+
+        if kind == SlruKind::Clog {
+            ensure!(block.len() == BLCKSZ as usize || block.len() == BLCKSZ as usize + 8);
+        } else {
+            ensure!(block.len() == BLCKSZ as usize);
+        }
```

Comment:
The `get_slru_keyspace` call, impl, and this piece of code that does special treatment for one variant are so far apart, I'd prefer this to be an exhaustive match, just to guard our future selves should there ever be a 4th kind of SLRU user.

Reply:
Done in [841aa9b](https://github.com/neondatabase/neon/pull/6372/commits/841aa9b0f0c5797ec060c415f1bf458f6c1770d2)